경북대 BE_신성민 4주차 과제 (1단계)

anaconda77 opened the pull request:
안녕하세요 경북대 BE 신성민 학생입니다. 4주차 과제 step1 제출드립니다.

- 월요일 수업 시간에 Mockito 라이브러리를 활용하여 계층별 단위 테스트를 하는 것을 학습하여서, 이번 과제에 Controller/Service 각 영역에 Mockito을 활용한 단위 테스트를 해보았습니다. 
mock을 도입하여 테스팅 해보는 것이 처음인지라 정확하게 작성하였는지 잘 모르겠습니다..어디까지 검증해야할지와 무엇을 검증할 것인가를 정하기가 어려웠던 것 같아요. 테스트 코드 위주로 피드백 해주시면 감사하겠습니다..!
`CategorySeviceTest.java`
`ProductServiceTest.java`
`CategoryApiControllerTest.java`

## 과제를 수행하면서 느낀 점

- 계층별 단위 테스트를 수행할때 검증하고자하는 메소드의 반환값이 있으면 이를 토대로 검증하기가 쉬웠지만, 반환값이 없는 void형인 경우 무엇을 검증해야할까하는 어려움이 있었습니다. mockito.then()을 활용하여 의존 객체의 특정 메소드가 실행되었는지를 검증하게 했지만, 단순히 실행 여부만 알려줄 뿐 요청한 작업을 정확하게 처리하는지는 알 수가 없어서 이게 맞는가 하는 생각이 들었습니다.
- 하지만 mock 기반 테스트를 여럿 수행하면서 제가 느낀 점은 의존 객체에서 반환되는 값 혹은 처리된 작업이 정확한지에 대해 집착할 필요는 없다는 생각이 들었습니다. 테스팅에 mock을 도입하는 목적이 테스팅 객체 자체의 검증에 최대한 집중하고, 의존 객체의 정상적 동작을 가정함으로써 의존 객체를 실제로 동작시키지 않고 빠른 테스팅을 수행할 수 있게 하기 위함이라고 생각하게 되었습니다. 의존 객체에 대한 단위 테스팅이 이미 정상적으로 수행되었다면 이러한 가정을 해도 무리 없지 않을까..? 하는 생각이 들었습니다.

anaconda77 committed:
feat: Category 모델 생성

anaconda77 committed:
feat: Product Category 연관관계 설정

anaconda77 committed:
feat: 데모 데이터 설정

anaconda77 committed:
feat: CategoryRepo 생성

anaconda77 committed:
feat: CategoryService 생성

anaconda77 committed:
refactor: 도메인 요구사항 변경

anaconda77 committed:
feat: category 생성시 이름 중복 검사

anaconda77 committed:
test: CategoryService 테스트 작성

anaconda77 committed:
feat: ProductService 카테고리 삽입/변경 기능 추가

anaconda77 committed:
test: ProductService 테스트 작성

anaconda77 committed:
feat: ProductController 카테고리 복합 요청 기능 추가

anaconda77 committed:
refactor: e2e 테스트 명칭 변경

anaconda77 committed:
feat: CategoryApiController 작성

anaconda77 committed:
refactor: PagingTest 별도의 테스트 케이스로 할당

anaconda77 committed:
test: CategoryApiController 테스트 작성

anaconda77 committed:
refactor: 요청/응답 파라미터 언더스코어 표기

anaconda77 committed:
feat: 카테고리 API 관리자 권한 인증 추가

lsm7179 commented on src/main/java/gift/model/Category.java:
```diff
@@ -0,0 +1,60 @@
+        columnNames = {"name"}
+    )
+})
+public class Category extends BaseEntity {
```
BaseEntity 잘 만들어주셨네요.

lsm7179 commented on src/main/java/gift/model/Category.java:
```diff
@@ -0,0 +1,60 @@
+        this.name = name;
+    }
+
+    public static void validateName(String name) {
+        if (name == null || name.isEmpty() || name.length() > 20) {
+            throw new InputException("이름을 1~20자 사이로 입력해주세요");
+        }
+    }
```
static으로 만들지 않아도 될거같습니다~

lsm7179 commented on src/main/java/gift/model/Member.java:
```diff
@@ -42,13 +42,16 @@ public Member(String email, String password) {
         this.role = role;
     }
 
+    public void makeAdminMember() {
+        this.role = Role.ROLE_ADMIN;
+    }
+
```
좋습니다.!

lsm7179 commented on src/main/java/gift/response/ProductResponse.java:
```diff
@@ -1,13 +1,20 @@
             product.getPrice(),
-            product.getImageUrl());
+            product.getImageUrl(),
+            product.getCategory().getName()
```
우리의 코드가 디미터의 법칙을 준수할수있도록 수정해보세요.
다른곳도 있다면 같이 확인해주세요~
https://mangkyu.tistory.com/147

anaconda77 replied:
Category 도메인에서 카테고리의 이름이 저장되는 위치나 방식이 변화될 수 있으므로, 결합력을 낮추기 위해 일종의 캡슐화된 메소드를 바라보도록 하는 것인가요?

lsm7179 replied:
네 맞습니다.
이렇게 되면 어떤 이점이 있을까요?
 productResponse는 카테고리의 내부 필드 값을 몰라도 되는 이점이 생깁니다.

lsm7179 commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -33,15 +39,24 @@ public Product getProduct(Long id) {
 
     @Transactional
-    public void editProduct(Long id, String name, Integer price, String imageUrl) {
+    public void editProduct(Long id, String name, Integer price, String imageUrl,
```
카테고리에서는 update로 메소드명이 되어있던데 둘중에 하나 동일하게 사용하는것이 가독성에 좋을거 같습니다.

lsm7179 commented on src/test/java/gift/service/CategoryServiceTest.java:
```diff
@@ -0,0 +1,158 @@
+import org.mockito.Mockito;
+import org.springframework.test.context.ActiveProfiles;
+
+@ActiveProfiles("test")
+class CategoryServiceTest {
+
+    private CategoryService categoryService;
+    private CategoryRepository categoryRepository = mock(CategoryRepository.class);
+
```
spring에서 제공하는 mockito 어노테이션을 활용해보아요.
https://mangkyu.tistory.com/145

lsm7179 reviewed (changes requested):
성민님 1단계 잘 해주셨습니다.
간단한 피드백 남겼는데 반영 후 재요청해주세요.

anaconda77 committed:
refactor: step1 1차 피드백 반영 및 리팩토링

- static 표기 삭제
- 디미터의 법칙 고려하여 메소드 레벨 추상화
- service 계층 메소드명 표기 통일
- 스프링 mockito 애노테이션 적용
- 기타 테스팅 리팩토링

anaconda77 commented:
오전에 피드백 주신 부분 수정해서 커밋 올려놓았습니다!

lsm7179 commented:
> 안녕하세요 경북대 BE 신성민 학생입니다. 4주차 과제 step1 제출드립니다.
> 
> * 월요일 수업 시간에 Mockito 라이브러리를 활용하여 계층별 단위 테스트를 하는 것을 학습하여서, 이번 과제에 Controller/Service 각 영역에 Mockito을 활용한 단위 테스트를 해보았습니다.
>   mock을 도입하여 테스팅 해보는 것이 처음인지라 정확하게 작성하였는지 잘 모르겠습니다..어디까지 검증해야할지와 무엇을 검증할 것인가를 정하기가 어려웠던 것 같아요. 테스트 코드 위주로 피드백 해주시면 감사하겠습니다..!
>   `CategorySeviceTest.java`
>   `ProductServiceTest.java`
>   `CategoryApiControllerTest.java`
> 
> ## 과제를 수행하면서 느낀 점
> * 계층별 단위 테스트를 수행할때 검증하고자하는 메소드의 반환값이 있으면 이를 토대로 검증하기가 쉬웠지만, 반환값이 없는 void형인 경우 무엇을 검증해야할까하는 어려움이 있었습니다. mockito.then()을 활용하여 의존 객체의 특정 메소드가 실행되었는지를 검증하게 했지만, 단순히 실행 여부만 알려줄 뿐 요청한 작업을 정확하게 처리하는지는 알 수가 없어서 이게 맞는가 하는 생각이 들었습니다.
> * 하지만 mock 기반 테스트를 여럿 수행하면서 제가 느낀 점은 의존 객체에서 반환되는 값 혹은 처리된 작업이 정확한지에 대해 집착할 필요는 없다는 생각이 들었습니다. 테스팅에 mock을 도입하는 목적이 테스팅 객체 자체의 검증에 최대한 집중하고, 의존 객체의 정상적 동작을 가정함으로써 의존 객체를 실제로 동작시키지 않고 빠른 테스팅을 수행할 수 있게 하기 위함이라고 생각하게 되었습니다. 의존 객체에 대한 단위 테스팅이 이미 정상적으로 수행되었다면 이러한 가정을 해도 무리 없지 않을까..? 하는 생각이 들었습니다.

좋은 고민을 해주셨네요~!
계층별 어떤 것을 검증할 것인가에 대해서 생각해보면 좋은데요.
우아한형제들 아티클을 한번 보면 좋을거 같네요.
https://techblog.woowahan.com/14874/

또한 통합테스트할때 mockbean을 이용하지 않으면 실제 디비로 저장되어 검증 할 수가 있습니다.
저같은 경우 테스트컨테이너를 이용하여 테스트시에 도커로 디비환경을 구축하여 통합테스트를 진행한 적도 있습니다.
https://testcontainers.com/
또한 springboot 3.1 부터 테스트컨테이너로 개발환경구축을 할 수 있다고 합니다.
https://spring.io/blog/2023/06/23/improved-testcontainers-support-in-spring-boot-3-1

런던파와 고전파의 대해서도 공부해보면 좋을거 같습니다. https://product.kyobobook.co.kr/detail/S000001805070

lsm7179 commented on src/main/java/gift/model/Category.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public void validateName(String name) {
+        if (name == null || name.isEmpty() || name.length() > 20) {
```
null 체크는 `Objects.isNull` 도 사용 할 수있습니다~

lsm7179 commented on src/test/java/gift/e2e/ProductApiTest.java:
```diff
@@ -40,7 +37,7 @@
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
 @ActiveProfiles("test")
-class ProductApiControllerTest {
+class ProductApiTest {
```
이름을 수정한 이유가 있나요?

anaconda77 replied:
해당 테스트가 컨트롤러에서 db까지 모든 과정을 테스팅하는 e2e 테스트에 가깝다고 생각하여서, 컨트롤러만 한정하여 테스트하는 파일과 이름을 구분짓기 위해 수정하였습니다.

lsm7179 reviewed (approved):
성민님 과제 잘 진행해주셨습니다.
테스트 코드가 꼼꼼한게 눈에 띄네요👍
수고하셨습니다~

anaconda77 commented:
멘토님 이번 과제와 별개로 Validation에 대한 질문 하나 드려도 괜찮을까요?
제 코드를 살펴보니 Entity에 들어갈 값에 대한 Validation이 Controller 영역에서 DTO를 생성할 때와, Service 영역에서 update()를 호출할 때에 수행되는 것을 확인하였습니다. 
```
public record ProductAddRequest(
    @NotBlank(message = "상품명을 입력해주세요.")
    @Length(min=1, max=15, message = "1~15자 사이로 입력해주세요.")
    @Pattern(regexp = "^[a-zA-Zㄱ-ㅎㅏ-ㅣ가-힣0-9()+&/_ \\[\\]-]*$", message = "상품명에 특수 문자는 '(, ), [, ], +, -, &, /, -' 만 입력 가능합니다.")
    @NoKakao
    String name,
    @NotNull(message = "가격을 입력해주세요.")
    @Range(min = 100, max = 1000000, message = "상품 가격은 100원~1,000,000원 사이여야 합니다.")
    Integer price,
    @NotBlank(message = "이미지 주소를 입력해주세요.")
    @Pattern(regexp = "^(https?)://[^ /$.?#].[^ ]*$", message = "올바른 url이 아닙니다.")
    @JsonProperty(value = "image_url")
    String imageUrl,
    @NotBlank(message = "카테고리를 입력해주세요.")
    @JsonProperty(value = "category_name")
    String categoryName
)
```
문제는 Entity를 생성할때 생성자를 통해 값을 Entity에 직접 주입해주게 되는데, 여기에는 Validation이 구현되어 있지 않아 혹여나 잘못된 값을 들어가버려도 이를 잡아내지 못할 것 같습니다.
```
public Product(Long id, String name, Integer price, String imageUrl, Category category) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.imageUrl = imageUrl;
        this.category = category;
    }
```

비록 지금 과제와 같이 한 사람이 모든 계층을 구현하는 경우에는 이런 오류를 발생시킬 확률이 낮지만, 여러 명이서 계층을 분리하여 협업하는 경우 Entity 자체에 대한 Validation이 누락되었을때 문제가 충분히 발생할 수 있을 것 같습니다.

Entity에 대한 자체 Validation은 어떻게 수행해주면 좋을까요? 생성자의 각 파라미터에 Validation 애노테이션을 붙여야 할까요..? 이렇게 되면 오류가 발생할 경우 에러는 어디에서 처리를 해주어야 할지 잘 모르겠습니다..!
